tools/swift-inspect/Sources/SwiftInspectLinux/SymbolCache.swift

```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LinuxSystemHeaders
+
+public class SymbolCache {
+  public typealias SymbolRange = (start: UInt64, end: UInt64)
+  public typealias SymbolInfo = (start: UInt64, end: UInt64, module: String, name: String)
+  public typealias SymbolLookup = [String: [String: SymbolRange]]
+
+  public let symbolLookup: SymbolLookup
+  let linkMap: LinkMap
+
+  // an array of all symbols sorted by their start address
+  lazy var sortedAddressLookup: [SymbolInfo] = {
+    var addressLookup: [SymbolInfo] = []
+    for (module, symbols) in self.symbolLookup {
+      for (name, (start, end)) in symbols {
+        addressLookup.append((start: start, end: end, module: module, name: name))
+      }
+    }
+    addressLookup.sort { $0.start < $1.start }
+    return addressLookup
+  }()
+
+  public init(for process: Process) throws {
+    self.linkMap = try LinkMap(for: process)
+    var symbolLookup: SymbolLookup = [:]
+    for linkMapEntry in linkMap.entries {
+      guard FileManager.default.fileExists(atPath: linkMapEntry.moduleName) else { continue }
```

Comment:
This feels like a TOCTOU issue.

Reply:
I suppose it is. I added this check so we don't try to open the "[vdso]" entry in the linkmap, but I could just as easily detect and skip that OR catch a not found exception from `try ElfFile(...)`.